ckanext/unaids/auth_logic.py

```diff
@@ -0,0 +1,155 @@
+import logging
+
+from ckantoolkit import config
+import json
+
+from repoze.who.interfaces import IChallengeDecider
+from six.moves.urllib.request import urlopen
+
+from flask import _request_ctx_stack, Response
+from jose import jwt
+from zope.interface import directlyProvides
+
+from ckan.common import request, g
+from ckan.logic import ActionError
+from ckan.model import Session, User, ApiToken
+
+AUTH0_DOMAIN = config.get('ckanext.unaids.auth0_domain')
+API_AUDIENCE = config.get('ckanext.unaids.oauth2_api_audience')
+REQUIRED_SCOPE = config.get('ckanext.unaids.oauth2_required_scope')
+ALGORITHMS = ["RS256"]
+
+log = logging.getLogger()
+
+class OAuth2AuthenticationError(ActionError):
+    pass
+
+
+class OAuth2AuthorizationError(ActionError):
+    pass
+
+
+def json_response_omitting_challenge_decider(environ, status, headers):
+    h_dict = dict(headers)
+    ct = h_dict.get('Content-Type')
+    if ct is not None:
+        return not ct.startswith('text/json') and status.startswith('401 ')
+
+    return status.startswith('401 ')
+
+
+directlyProvides(json_response_omitting_challenge_decider, IChallengeDecider)
+
+
+# based on work from https://auth0.com/docs/quickstart/backend/python/01-authorization
+def access_token_present_and_valid_and_user_authorized():
+    token = request.headers.get('Authorization', '')
+    if token:
+        access_token = validate_and_decode_token(token)
+        verify_required_scope(access_token)
+        subject = access_token['sub']
+
+        if not subject.startswith("auth0|"):
```

Comment:
If we want to allow other identity providers (e.g. 'google-oauth2') we can change it into a variable. Or remove this check at all.